### PR TITLE
Add example drugban py g

### PR DESCRIPTION
**Pull request by jiayang-zhang**
Fixes #{issue_number}.

### Description
Integrated Peizhen's DrugBAN to PyKale's workflow, with DGL being replaced by PyG.

1. Added a new folder for DrugBAN in the `examples` directory as a starting point for using this model within PyKale.
2. Added `ban.py` to `kale.embed` for DrugBAN and BANLayer.
3. Added `drugban_datasets.py` to `kale.loaddata` for loading datasets specific to DrugBAN.
4. Added `drugban_trainer` to `kale.pipeline`.
5. Created a `Discriminator` class in `kale.predict.class_domain_nets` to support domain adaptation.
6. Added `binary_cross_entropy` function in `kale.evaluate.metrics` for loss calculation.

### Status
**Ready**

If you've been given access to pykale with role Triage or above, on the right (delete these after selection):

- Select **one most appropriate** label.
- **When** your pull request is **ready** for review, select a reviewer. Use the suggested one if unsure.

If you cannot see an option to select a reviewer/label, that means one maintainer will get notified upon your pull request and then select for you.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] In-line docstrings updated.
- [ ] [Source for documentation at `docs`](https://github.com/pykale/pykale/tree/main/docs/source) manually updated for new API.

### CODE REVIEW

File: kale/loaddata/molecular_datasets.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+        Parameters:
+        -----------
+        index : int
+            The index of the sample to retrieve.
+
+        Returns:
+        --------
+        v_drug : DGLGraph
+            A tensor representing the drug molecule, with node features and optional virtual nodes.
```
[the remark this answers is not in the document]
Comment (jiayang-zhang) [reply kale/loaddata/molecular_datasets.py]: The suggestion has been taken.

File: kale/pipeline/drugban_trainer.py [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+    def train_epoch(self):
+        """
+        Perform a single epoch of training.
+
+        This method trains the model for one full pass (epoch) over the training dataset.
+        """
+        self.model.train()
+        loss_epoch = 0
+        num_batches = len(self.train_dataloader)
+        for i, (v_drug, v_protein, labels) in enumerate(tqdm(self.train_dataloader)):
```
Comment (shuo-zhou) [inline kale/pipeline/drugban_trainer.py:351]: What does v stand for?

File: kale/pipeline/drugban_trainer.py [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+        if dataloader == "test":
+            data_loader = self.test_dataloader
+        elif dataloader == "valid":
+            data_loader = self.valid_dataloader
+        else:
+            raise ValueError(f"Error key value {dataloader}")
+        num_batches = len(data_loader)
+        with torch.no_grad():
+            self.model.eval()
+            for i, (v_drug, v_protein, labels) in enumerate(data_loader):
```
Comment (shuo-zhou) [inline kale/pipeline/drugban_trainer.py:531]: use _ if the i is not used in the loop

File: kale/pipeline/drugban_trainer.py [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+        self.optim_da = None
+        self.device = device
+        self.epochs = config["SOLVER"]["MAX_EPOCH"]
+        self.current_epoch = 0
+        self.train_dataloader = train_dataloader
+        self.valid_dataloader = valid_dataloader
+        self.test_dataloader = test_dataloader
+        self.is_da = config["DA"]["USE"]
+        self.alpha = alpha
+        self.n_class = config["DECODER"]["BINARY"]
```
Comment (shuo-zhou) [inline kale/pipeline/drugban_trainer.py:79]: `num_classes`, keep consistent with existing naming conventions.

File: kale/pipeline/drugban_trainer.py [hunk 5]
```diff
@@ -0,0 +1,570 @@
...
+
+        # ---- setup domain adaption model and optimizer ----
+        if self.is_da:  # If domain adaptation is used
+            if config["DA"]["RANDOM_LAYER"]:
+                domain_dmm = Discriminator(
```
Comment (shuo-zhou) [inline kale/pipeline/drugban_trainer.py:84]: `domain_discriminator` may be more readable

File: kale/pipeline/drugban_trainer.py [hunk 6]
```diff
@@ -0,0 +1,570 @@
...
+            auroc, auprc, valid_loss = self.test(dataloader="valid")
+            if self.experiment:
+                self.experiment.log_metric("valid_epoch model loss", valid_loss, epoch=self.current_epoch)
+                self.experiment.log_metric("valid_epoch auroc", auroc, epoch=self.current_epoch)
+                self.experiment.log_metric("valid_epoch auprc", auprc, epoch=self.current_epoch)
+            valid_lst = ["epoch " + str(self.current_epoch)] + [f"{x:.4f}" for x in [auroc, auprc, valid_loss]]
+            self.valid_table.add_row(valid_lst)
+            self.valid_loss_epoch.append(valid_loss)
+            self.valid_auroc_epoch.append(auroc)
+            if auroc >= self.best_auroc:
```
Comment (shuo-zhou) [inline kale/pipeline/drugban_trainer.py:221]: I suggest use a dictionary for saving all results

File: kale/pipeline/drugban_trainer.py [hunk 7]
```diff
@@ -0,0 +1,570 @@
...
+
+        Args:
+        logits (torch.Tensor): The logits output from the model.
+
+        Returns:
+        torch.Tensor: The computed entropy-based weights for each sample in the batch.
+        """
+        entropy = entropy_logits(logits)
+        entropy = ReverseLayerF.apply(entropy, self.alpha)
+        entropy_w = 1.0 + torch.exp(-entropy)
```
Comment (shuo-zhou) [inline kale/pipeline/drugban_trainer.py:339]: `entropy_weights` could be better

File: kale/pipeline/drugban_trainer.py [hunk 8]
```diff
@@ -0,0 +1,570 @@
...
+                        if self.random_layer:
+                            random_out_t = self.random_layer.forward(feature_t)
+                            adv_output_tgt_score = self.domain_dmm(random_out_t)
+                        else:
+                            adv_output_tgt_score = self.domain_dmm(feature_t)
+
+                    if self.use_da_entropy:
+                        entropy_src = self._compute_entropy_weights(score)
+                        entropy_tgt = self._compute_entropy_weights(t_score)
+                        src_weight = entropy_src / torch.sum(entropy_src)
```
Comment (shuo-zhou) [inline kale/pipeline/drugban_trainer.py:452]: Be consistent for all src, tgt, better to keep them at the end. Check all

File: setup.py [hunk 9]
```diff
@@ -57,6 +57,8 @@
     "yacs>=0.1.7",
     "pwlf",
     "xlsxwriter",
+    "prettytable",
```
Comment (shuo-zhou) [inline setup.py:60]: Not only examples need them. The core functions need these libraries, should be listed in core dependencies

File: kale/evaluate/metrics.py [hunk 10]
```diff
@@ -45,6 +45,21 @@ def cross_entropy_logits(output, target, weights=None):
     return loss, is_correct
 
 
+def binary_cross_entropy(output, target):
```
Comment (shuo-zhou) [inline kale/evaluate/metrics.py:48]: If `cross_entropy_logits` above is not applicable, please consider merge these two functions.

File: kale/pipeline/drugban_trainer.py [hunk 11]
```diff
@@ -0,0 +1,570 @@
...
+from torch.nn import functional as F
+from tqdm import tqdm
+
+from kale.embed.ban import RandomLayer
+from kale.evaluate.metrics import binary_cross_entropy, cross_entropy_logits, entropy_logits
+from kale.pipeline.domain_adapter import GradReverse as ReverseLayerF
+from kale.predict.class_domain_nets import Discriminator
+
+
+class Trainer(object):
```
Comment (shuo-zhou) [inline kale/pipeline/drugban_trainer.py:37]: The class name should be more specific, e.g., `DrugBANTrainer`. Usually deep learning trainer in PyKale should inherit from pl.LightningModule. Learn from examples [here](https://github.com/pykale/pykale/blob/2bb887077ea54a14d12ea6a03255413773e9355f/kale/pipeline/domain_adapter.py#L171).

File: kale/pipeline/drugban_trainer.py [hunk 12]
```diff
@@ -0,0 +1,570 @@
...
+        config (dict): Configuration dictionary containing model, DA, and training settings.
+
+
+    """
+
+    def __init__(
+        self,
+        model,
+        device,
+        train_dataloader,
```
Comment (shuo-zhou) [inline kale/pipeline/drugban_trainer.py:61]: Better to use dataset as input instead of separate dataloaders. Learn from other [pykale trainers](https://github.com/pykale/pykale/blob/2bb887077ea54a14d12ea6a03255413773e9355f/kale/pipeline/domain_adapter.py#L219).

**Review by shuo-zhou — CHANGES_REQUESTED** [review]

File: kale/pipeline/drugban_trainer.py (hunk 11 above)
Comment (jiayang-zhang) [reply kale/pipeline/drugban_trainer.py:37]: This is a PyTorch non-lightning implementation. The lightning ver. can be adapated once this PyTorch ver. is merged.

File: kale/pipeline/drugban_trainer.py [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+
+
+class DrugbanTrainer(pl.LightningModule):
+    """
+    Pytorch Lightning DrugBAN Trainer class
+
+    The DrugBAN Trainer class encapsulates the logic for training a DrugBAN model, optionally with domain adaptation (DA),
+    and manages the optimization, logging, and saving of the best-performing model.
+
+    Args:
```
Comment (shuo-zhou) [inline kale/pipeline/drugban_trainer.py:34]: docstrings for all args

File: kale/pipeline/drugban_trainer.py [hunk 14]
```diff
@@ -0,0 +1,350 @@
...
+        self.use_da_entropy = use_da_entropy
+
+        # ---- setup domain adaption model and optimizer ----
+        if self.is_da:  # If domain adaptation is used
+            # set up discriminator
+            if da_random_layer:
+                # Initialize the Discriminator with an input size from the random dimension specified in the config
+                # self.domain_discriminator = Discriminator(input_size=da_random_dim, n_class=self.num_classes)
+                self.domain_discriminator = DomainNetSmallImage(
+                    input_size=da_random_dim,
```
Comment (shuo-zhou) [inline kale/pipeline/drugban_trainer.py]: Only need to specify different input size.

File: kale/pipeline/drugban_trainer.py [hunk 15]
```diff
@@ -0,0 +1,350 @@
...
+        self.test_metrics = metrics.clone(prefix="test_")
+
+    def configure_optimizers(self):
+        """
+        Configure the optimizers for the DrugBAN model and the Domain Discriminator model if DA is used.
+
+        """
+        if not self.is_da:  # If domain adaptation is not used
+            # Initialize the optimizer for the DrugBAN model
+            opt = torch.optim.Adam(self.model.parameters(), lr=self.solver_lr)
```
Comment (shuo-zhou) [inline kale/pipeline/drugban_trainer.py:159]: Not sure if using two separate optimizers works well in lightning. This is the key difference between cross-domain test with and without DA

**Review by shuo-zhou — APPROVED** [review]
Approved to proceed with developing the notebook for the EMBC25 workshop. Assigned to @shuo-zhou for refactoring.
